lnwallet/commitment.go

```diff
@@ -835,9 +871,16 @@ func (cb *CommitmentBuilder) createUnsignedCommitmentTx(ourBalance,
 	}
 
 	// Sort the transactions according to the agreed upon canonical
-	// ordering. This lets us skip sending the entire transaction over,
-	// instead we'll just send signatures.
-	InPlaceCommitSort(commitTx, cltvs)
+	// ordering (which might be customized for custom channel types, but
+	// deterministic and both parties will arrive at the same result). This
+	// lets us skip sending the entire transaction over, instead we'll just
+	// send signatures.
+	commitSort := auxResult.CommitSortFunc.UnwrapOr(DefaultCommitSort)
+	err = commitSort(commitTx, cltvs, htlcIndexes)
```

Comment:
If there are multiple HTLCs with the same amount, same CLTV timeout and same payment hash (e.g. MPP shards with same split amount), they look identical on the BTC level (exact same script and amount), so the sort order of them is "fungible" (e.g. it doesn't matter which comes first in the transaction).
But on the Taproot Asset level it matters, because the actual asset leaves that are committed to each HTLC might be different. So we need to know exactly what output index a specific HTLC is located at in the final on-chain transaction. Therefore we need to take the HTLC index as an additional sorting indicator.

I didn't want to put any asset specific comments into the lnd code. But I added a bit more rationale to the documentation of the `CommitSortFunc` comment.

Reply:
I must be missing something: [Bolt 3](https://github.com/lightning/bolts/blob/master/03-transactions.md#transaction-output-ordering) specifies that `scriptpubkey` is part of the sort order and afaik the tapscript root affects that, doesn't it?

Reply:
Yes, you are correct. We're basically sorting by amount and `scriptpubkey` and the tapscript root affects that. But because the actual custom channel tapscript leaf depends on the output index and the output index changes depending on the leaf, we have a potentially infinite recursion.
So we decided to break with Bolt3 and just do a single iteration of sort using the fields mentioned above, then add the extra leaves even though that causes the transaction not to be BIP-069 sorted anymore.
IMO this is fine as both parties of custom channels will be able to deterministically do that. The only slight downside is that such transactions would be identifiable as custom channel transactions if they ever went on chain.